lib/pbench/test/unit/server/test_file_tree.py

```diff
@@ -358,29 +365,22 @@ def mock_rmtree(path: Path, ignore_errors=False):
             assert ignore_errors
             assert path == incoming / f"{tarball_name}.unpack"
 
-        def mock_run(args, **kwargs):
-            if args[0] == "tar":
-                return subprocess.CompletedProcess(
-                    args, returncode=0, stdout="Successfully Unpacked!", stderr=None
-                )
-            else:
-                assert args[0] == "find"
-                raise subprocess.TimeoutExpired("find", 43)
+        @staticmethod
+        def mock_run(command, dir_path, exception, dir_p):
+            verb = "find"
+            if command.startswith(verb):
+                raise subprocess.TimeoutExpired(verb, 43)
 
         with monkeypatch.context() as m:
             m.setattr(Path, "mkdir", lambda path, parents: None)
-            m.setattr("subprocess.run", mock_run)
+            m.setattr(Tarball, "subprocess_run", mock_run)
             m.setattr(shutil, "rmtree", mock_rmtree)
             m.setattr(Tarball, "__init__", TestFileTree.MockTarball.__init__)
             tb = Tarball(tar)
-            unpack_dir = incoming / f"{tarball_name}.unpack"
 
-            with pytest.raises(TarballModeChangeError) as exc:
+            with pytest.raises(Exception) as exc:
                 tb.unpack(incoming, results)
-            assert (
-                str(exc.value)
-                == f"An error occurred while changing file permissions of {unpack_dir}: Command 'find' timed out after 43 seconds"
-            )
+            assert str(exc.value) == "Command 'find' timed out after 43 seconds"
```

Comment:
This doesn't actually check what exception class instance was raised (ditto lines 347-349).  I _think_ part of the interface contract for `Tarball.unpack()` is that it may raise `TarballUnpackError` or `TarballModeChangeError`, so the unit tests should probably confirm that behavior.  (That is, if `Tarball.unpack()` raised some _other_ exception class instance, like a `subprocess.TimeoutExpired`, which happens to have that message text, that would be a violation of the contract, and this test wouldn't catch it.)

There are two ways to address this.  You can either add an assertion at line 383 which checks that the exception object in `exc` is an instance of `TarballModeChangeError`, or you can change the `pytest.raises()` call to expect `TarballModeChangeError` instead of `Exception`.  (And, you should do the analogous thing at line 347/349.)

Also, line 415.

Reply:
This error is being raised by the whole `subprocess_run` method and not `subprocess.run`. due to which when the error is being raised the exc is not an isnstance of `TarballUnpackError` or `TarballModeChangeError`. When we are specifically testing `subprocess.run` the errors that are being raised there are handled as they should have using `TarballUnpackError` or `TarballModeChangeError`.

Reply:
I don't think either of the two ways that are suggested will work here as the `exc` is not an instance of any (`TarballUnpackError` or `TarballModeChangeError`) and the same way if I add them in `pytest.raises()` it is not getting caught.

Reply:
OK, I was confused.  (And, I apologize for the trouble!)

We're mocking `Tarball.subprocess_run()`, and the mock function is raising `subprocess.TimeoutExpired(verb, 43)`, which is an interesting problem:  the real function should have raised either `TarballUnpackError` or `TarballModeChangeError`, because that is what `Tarball.unpack()` should be telling it to do.

So, I think our mock should be doing something similar.  Since the mock receives the `exception` parameter, it seems like it should be using that class in the `raise` (because that's what the real function should/would do).

The result of that should be that `exc` (which, I think is an [`ExceptionInfo`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.ExceptionInfo) object, not a subclass of `Exception`) should contain either a `TarballUnpackError` or `TarballModeChangeError` instance, depending on whether it was the `tar` or `find` command (emulation) which failed, respectively.  So, you should then be able to add an assertion like
```
            assert exc.type == TarballUnpackError
```
You might also want to consider using [`ExceptionInfo.match()`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.ExceptionInfo.match) to test the message text -- it provides the assertion for you, you just have to supply the text, and it will accept regex patterns if we need it to.